### DISCUSSION TITLE
Publish full JSON schema

### DISCUSSION THREAD

**Discussion by Radiergummi** [lowdefy/lowdefy#677] [Ideas and feature requests]
As discussed with @Gervwyk on Discord, it would be great to have a full JSON schema for Lowdefy available at some public URL, or preferably at [SchemaStore.org](https://www.schemastore.org/json/), which has a native integration with common IDEs (e.g. VS Code or IntelliJ).

That way, users had autocomplete for their `lowdefy.yaml` file out of the box, making development a much smoother experience. I'm not too deep into schema development, so I'm not sure what's necessary to get there, but I'm happy to help where I can!

**Comment by Gervwyk** [comment]
Hi @Radiergummi This would be a great addition. We've documented all blocks etc. using JSONSchema with this in mind. 

I've not created schemas for autocompletion in the past, but would be happy to help you experiment until we find a working solution.

I guess the starting point would be taking the current schemas and then converting them into a version that works for autocompletion. 
So we have schemas in a few places:

The main app schema: 
https://github.com/lowdefy/lowdefy/blob/main/packages/build/src/lowdefySchema.json

All block schemas for block setting are next to the blocks like here:
https://github.com/lowdefy/lowdefy/blob/main/packages/blocks/blocksAntd/src/blocks/Button/Button.json

For connections and requests aswell:
https://github.com/lowdefy/lowdefy/blob/main/packages/graphql/src/connections/MongoDBCollection/MongoDBCollectionSchema.json

So what we need to investigate is how we need to combine these schemas perhaps with a script into a schema we can publish to SchemaStore.org 

Also another consideration would be how to do autocompletion based on the block type, so that you see all properties for completion only related to the specific block type.

**Comment by Radiergummi** [comment]
By pointing the IDE to the main app schema like the following:
<img width="365" alt="Bildschirmfoto 2021-06-30 um 12 58 12" src="https://user-images.githubusercontent.com/6115429/123949424-d7992780-d9a2-11eb-8384-8a3653953ff2.png">

We get proper completion already, which is great:
<img width="172" alt="Bildschirmfoto 2021-06-30 um 12 57 34" src="https://user-images.githubusercontent.com/6115429/123949372-c7814800-d9a2-11eb-87d2-100231b4ee47.png">

I also think a schema building script will be the best way forward - something that analyzes all relevant schemas, and merges them into a single JSON schema definition for all of Lowdefy. From the documentation, it looks like there's a few options for conditionals:
See [json-schema.org/understanding-json-schema/reference/conditionals.html](https://json-schema.org/understanding-json-schema/reference/conditionals.html), or [this SO answer](https://stackoverflow.com/a/38781027/2532203) for a great summary.  
This way, the script could build the conditional logic required for block types, and we should end up with a super-flexible, updated-and-published-on-release, JSON schema.  
I'm not too familiar with the code base, but I'd assume there's a way to collect all JSON files in the main packages that contain schema definitions? In that case, (famous last words) _it doesn't sound too hard to map-reduce this into a valid JSON schema._

**Reply by Radiergummi** [reply]
On second thought, would it make sense to change the block schema to follow the JSON schema spec right away? Maybe that way the main schema could simply refer to them using `$ref` (see [here](https://json-schema.org/learn/getting-started-step-by-step#references-outside-the-schema)), and the publish CI step would simply flatten the schema to a single file?

**Reply by SamTolmay** [reply]
So maybe the blocks packages have an "index" schema, which reference all of the blocks in the package. So then the main schema references all the blocks packages, and the blocks packages reference their blocks.

Are we happy to reference the app schema from Github, or do we want to host those schemas on JSON schema store?

**Reply by Gervwyk** [reply]
I think once we have the details setup up, we would publish to the schema store on release, would be great if you can pull the schema version for the version of lowdefy define in your lowdefy.yaml. But this would be added later.

**Reply by Gervwyk** [reply]
I think both methods are viable options. My first approach would be to use a script since I think it might be easier to scale with complexity. But not sure. Since everything including docs is in this monorepo it would be reasonable think we can make this happen with ease. 

One thing that we would need to add is JSON schemas for operators - thinking of it now makes me wonder why we did not do this in the first place. Luckily, the operator schemas are mostly simple and documented well. So this should be quick. We would then convert the operator docs to read from the schemas aswell, like we do for blocks. This would be a nice improvement.

**Reply by Gervwyk** [reply]
The schema dependancies are would be nice to know which block properties to use. Just wondering, is there a way to run script that evaluates on autocomplete and then returns the schema for example?

**Reply by Radiergummi** [reply]
Kind of, using conditional logic:
> A relatively new addition to [JSON Schema (draft-07)](http://json-schema.org/draft-07/json-schema-release-notes.html) adds the `if`, `then` and `else` keywords. **_If_ the "foo" property equals "bar", _Then_ the "bar" property is required.**
> ```
> {
>   "type": "object",
>   "properties": {
>     "foo": { "type": "string" },
>     "bar": { "type": "string" }
>   },
>   "if": {
>     "properties": {
>       "foo": { "const": "bar" }
>     },
>     "required": ["foo"]
>   },
>   "then": { "required": ["bar"] }
> }

The above should match both - if foo is "bar", the bar property is required. Otherwise, bar isn't required.
```
{ "foo": "bar", "bar": "baz" }
{ "foo": "quz" }
```

So I guess this would allow defining different property schemas depending on the block type too!

**Comment by Gervwyk** [comment]
Another thing to consider here is the use of the _ref operator. Which might complicate things I think.

**Reply by SamTolmay** [reply]
This is true. The basic schema assumes all `_ref` operators are evaluated, and the block and connection schemas are based on when all runtime operators are evaluated.

But if we are only using the schema for autocomplete, it isn't a problem if schema validation fails, as long as the IDE gives good suggestions. Not sure how the IDE will "know" what part of the schema a referenced file is part of.

**Comment by Gervwyk** [comment]
Maybe one approach to consider that I think would be better and serve better results is to create a [Language Extention](https://code.visualstudio.com/api/language-extensions/programmatic-language-features#show-code-completion-proposals) instead of a complex JSONSchema. I've just scanned but perhaps this could be a better approach?

The logic could be as follow, from the cursor position, we just need to move up in the json structure within the same file to the first type or operator definition. Then you know which json schema to use in order to provide accurate autocompletion for the current object. You do not need the entire schema, you only need the schema of the object you are editing. This would also mean that _ref is not an issue since you only care about the file you are currently working on.

In terms of adding this, we could create an extension package which an publish places all the schema files on the cdn for the version, we are already doing it for blocks, then the package could just load and cache the schema files for the version of lowdefy you are working with. we probably do not have to update the extension for every new lowdefy version unless we add new features to the extension.   

I feel like I need to better my understanding of [language servers](https://code.visualstudio.com/api/language-extensions/language-server-extension-guide). 

Any experience with this or thoughts?
Does IntelliJ have a simular API for extensions?

**Reply by Gervwyk** [reply]
Ahh. Seems like I messed up the link. What I meant here is that we need to create a Language Extention for code completion:
https://code.visualstudio.com/api/language-extensions/programmatic-language-features#show-code-completion-proposals
It could be much easier, more powerful, and easier to maintain than the schema approach.

**Reply by Radiergummi** [reply]
I think LSP is a good way to provide tooling support and solve this issue. However, it feels like a project that has such a tight integration with JSON schema should be able to expose the full schema to the outside world. That allows other use cases too, like static analysis, validation, and (potentially) GUI editors, all independent of a development environment.  

Maybe both approaches would be useful? The schema for basic completion and integration with other tools, and the language extension for IDE code completion (I realize that's doubling the amount of work tho, so please don't take this for anything other than brainstorming 😉).

**Reply by Gervwyk** [reply]
I agree. It would make sense to do both. Although the schema would not work well for code autocomplete (because of the _ref operator) it would be very useful for a list of other things. 

Currently, we do validate the app config at build time, but only for the app schema and not on a block / operator / request level.

It's more work, but let's keep it in mind as we structure the project to not duplicate work. 

So to do the schema we just need to write a script which "collects" all schema files and assembles a single schema file?
Let's first see if we can compile a schema that also validates up to block level?

**Reply by SamTolmay** [reply]
The other thing to note is that the schema of the "built" does not strictly match the schema of the app configuration, as the build process does apply transformations to the configuration at build. This schema can also vary between different versions of the app, as it is internal to that app version. So there is never really a point in the app lifecycle that the app fully conforms to the schema.
